rust/connlib/tunnel/src/client.rs

```diff
@@ -733,6 +753,30 @@ impl RoleState for ClientState {
                 Poll::Pending => {}
             }
 
+            if self.refresh_dns_timer.poll_tick(cx).is_ready() {
```

Comment:
I've seen in a few places in connlib that we do `poll_ready` or `poll_*` things like this instead of the usual `timer.tick.await` inside a spawned task. What's the advantage? Is that to make it more independent of Tokio or something?

Reply:
The main benefit is that we can have mutable reference from within the `poll` call and therefore we don't need a `Mutex` or `Arc` to keep track of the state, having less room to contentions and deadlocks.

(cc @thomaseizinger if you want to add something)

Reply:
Yes! async/await is useful if you want to express a linear execution flow. Because an async fn gets converted into a `Future`, it becomes opaque once it runs within the runtime. In the current language design, the yield points generated by `.await` only suspend the underlying co-routine without an ability to specify a different resume argument. So to add more state or alter the co-routine (read: Future) otherwise, you need to use channels.

In connlib, we don't have linear execution flow but concurrent access to the state from the device, the socket and messages from the portal:

- Incoming data from the socket needs to be decrypted
- Incoming data on the device (the virtual network adapter) needs to be encrypted and sent
- Incoming messages from the portal need to alter the state

By hand-rolling the Future implementation using `poll`  functions, we can write code that has a different branching behaviour at the yield points: Normalling, `Poll::Pending` yields back to the executor. Instead, we can make progress on a _different_ future. For example, if the device is busy we can check for messages on the portal instead.

Overall, this allows us to have a single owner of the state, device and socket which makes it possible to use `&mut self` and handling borrowed data. The alternative is 3 tasks that use a lock around the state that they are competing for whenever a message comes in on the device, socket or from the portal.

Reply:
@ReactorScram Have a look at https://github.com/firezone/firezone/pull/2965. It deals with a much smaller scope of this problem because it leaves out DNS, resource policies, device and portal messages.

The code within the test binaries could be cleaned up a bit more that is essentially what connlib will look like at some point. Just imagine that the loop does not only poll the socket but also the device and output from one is input of the other.